### ISSUE TITLE
add option to local-css to prefix minified classnames

### ISSUE THREAD

**Issue by mpech** [evanw/esbuild#3484]
link of original issue which was resolved: https://github.com/evanw/esbuild/issues/3295

Our problem is the following
- we have library A bundled by esbuild
- client B which depends on library A, also bundled **after** by esbuild
=> when B builds, B's stylesheet clashes with A's stylesheet, because we do not specify A as entrypoint, we just get the built files from A

What we would appreciate
- enable customizing classname like 'A[classnamehash]' (this would be the best, we could prefix our classes by libraryname)
- either specify A's stylesheet when building B to esbuild (it may be impractical, because if B is actually not a client but a library too, and if B uses A as peerdeps, while bundling B we will avoid clashes from A, but when building the client, maybe the classnames generated of A will have changed by then, which forces to have strict versionning)

**Comment by mhsdesign** [comment]
Yes i agree. I currently have encountered the same problem. Im building a javascript plugin and it seems that another plugin already uses esbuild which causes css classes to overlap randomly. A simple prefix like `my-vendor` would be perfect.

lightningcss  also allows to declare a prefix via their "pattern" option: https://lightningcss.dev/css-modules.html#custom-naming-patterns

**Comment by zouloux** [comment]
Hi, this is also an issue on some projects here.
I didn't find a way to intercept renaming through a plugin.
If anybody has a temporary solution, it would be helpful. In the meanwhile, we will disable the minify option.

**Comment by mhsdesign** [comment]
You can use another plugin for css modules - like mine ^^ https://github.com/mhsdesign/esbuild-plugin-lightningcss-modules its based on lightningcss

**Comment by markplindsay** [comment]
@evanw, my team develops a browser extension, and I recently switched us to esbuild for the dramatic speed improvement. Thank you very much for writing this software.

Our extension coexists in the DOM on many websites, and it is subject to these sites' global styling. I was troubleshooting a visual bug this week, and I noticed that Reddit uses a classname simply named `.a`. (Perhaps it is even generated by esbuild.)

Unfortunately, this conflicts with an `.a` classname generated by our esbuild during local-css's minification process,  and it messes up our extension's styling when it is rendered on Reddit.

We're currently working around this by not minifying our extension, but introducing an optional prefix for generated local-css classnames as configuration would solve this problem for us 100%.

**Comment by kachurun** [comment]
I have the same problem. I'm creating a powerful HTML element for my website builder, and every single HTML element has its own bundle. Now, when two or more elements with CSS modules are involved, it causes this issue.

**Comment by chowey** [comment]
I just have a site with pre-existing CSS. Someone in the distant past decided to use a short class name. Now it conflicts with the minified class names from esbuild.

So I'd also like to see an option in esbuild to prefix class names.

**Comment by chowey** [comment]
I took a stab at this with #3914. It's working for me locally.

I called the new setting `local-css-prefix`.

**Comment by ziemkowski** [comment]
#3914 would be helpful.

Using CSS Modules with Rollup/LightningCSS has encouraged the team to go back to using simple class names like `.btn`, `.contianer`, `.link`, `.row`, etc., which are defined in most CSS framework out there and collide if esbuild-generated local CSS classnames can't be made unique via prefix or hash.

Unfortunately, the CSS Module plugins I've tried so far to get around this break down with `splitting: true`.

**Comment by chowey** [comment]
Here's a workaround that is working for me:

Simply define the conflicting classname in a global CSS file that is bundled with esbuild. Then esbuild won't use it for local CSS classnames.

---

[For example](https://esbuild.github.io/try/#YgAwLjI0LjIALS1idW5kbGUgLS1taW5pZnkgLS1vdXRkaXI9LgBlAGVudHJ5LmpzAGltcG9ydCAiLi9sb2NhbC5tb2R1bGUuY3NzIjsAAGxvY2FsLm1vZHVsZS5jc3MALm15TG9jYWxDbGFzc05hbWUgewogIGNvbG9yOiByZWQ7Cn0), here esbuild converts `myLocalClassName` to `o`:

```css
/* local.module.css */
.myLocalClassName {
  color: red;
}
```

```css
/* minified */
.o{color:red}
```

But lets say `o` conflicts with an existing classname on the site. You can force esbuild to pick another classname it by defining `o` in a global stylesheet and bundling that stylesheet.

[For example](https://esbuild.github.io/try/#YgAwLjI0LjIALS1idW5kbGUgLS1taW5pZnkgLS1vdXRkaXI9LgBlAGVudHJ5LmpzAGltcG9ydCAiLi9sb2NhbC5tb2R1bGUuY3NzIjsKaW1wb3J0ICIuL2dsb2JhbC5jc3MiAABsb2NhbC5tb2R1bGUuY3NzAC5teUxvY2FsQ2xhc3NOYW1lIHsKICBjb2xvcjogcmVkOwp9AABnbG9iYWwuY3NzAC5vIHsKICByZXNlcnZlZDogdHJ1ZTsKfQ):

```css
/* local.module.css */
.myLocalClassName {
  color: red;
}

/* global.css */
.o {
  reserved: true;
}
```

```css
/* minified */
.e{color:red}.o{reserved:true}
```

Now esbuild picked `e`. There is nothing special about the `reserved: true` CSS rule except it doesn't mean anything so it will not affect the styling of the `o` class.

**Comment by nicksrandall** [comment]
I've run into the issue of colliding class names as well. 

I was imagining there could be an option similar to [asset-names](https://esbuild.github.io/api/#asset-names) but for class name generation. We could use that to provide a prefix or hash to the generated classes.

**Comment by nicksrandall** [comment]
I've run into this again today... I would really love for esbuild to support (component) libraries and not just end-user apps.

@evanw I'd be happy to implement something if you could give some direction on the API that you would prefer.

**Comment by OEvgeny** [comment]
Just for everyone interested in a working solution while this is being addressed, we were able to use the [unplugin/unplugin-lightningcss](https://github.com/unplugin/unplugin-lightningcss) as a plugin for building CSS with CSS Modules without much effort.

I think esbuild's CSS support is still rather experimental. It's a huge pain that two separate bundles can't be used on the same page.

**Comment by joelmoss** [comment]
> I was imagining there could be an option similar to [asset-names](https://esbuild.github.io/api/#asset-names) but for class name generation. We could use that to provide a prefix or hash to the generated classes.

I would that this would be the most flexible and powerful option, and would be extremely useful.

**Comment by mhsdesign** [comment]
Okay its not really worth the bits send over the wire but ill still post my hack :D

I _have_ build a proper css modules integration already with https://github.com/evanw/esbuild/issues/3484#issuecomment-1807780793 but i like to use esbuild natively because its faster and less to setup. Still im troubled by the duplicated css class names which happens very often when writing javascript and css plugins for another system say a CMS.

I hope that any of the two prs resolve this soon but for now ill just do a dirty string replace on the result to ensure unique prefixes.

My code below only works for all css module imports that are formed like `import style from './style.module.css';` (which is the convention i follow) and i then replace esbuilds `.style_buttonA` with `.myPrefix_buttonA` 

<details><summary>My Funny code</summary>
<p>

```js
plugins: [
    {
        // HACK for https://github.com/evanw/esbuild/issues/3484
        name: 'prefix-local-css',
        setup(build) {
            // insert replacement
            const replace = 'xxxx_';
            const search = 'style_';

            if (replace.length !== search.length) {
                throw new Error(`"${replace}" and "${search}" should have the same length.`)
            }

            if (build.initialOptions.minify) {
                build.initialOptions.minify = false;
                // don't minify CSS class names but keep them in format of ".style_someButton" instead of ".a" when used like `import style from './style.module.css';`
                build.initialOptions.minifyIdentifiers = false;
                // this can be minified
                build.initialOptions.minifySyntax = true;
                build.initialOptions.minifyWhitespace = true;
            }

            build.initialOptions.write = false;
            build.onEnd(result => {
                const nodeFs = require('node:fs');
                const nodePath = require('path');

                for (const outputFile of result.outputFiles) {
                    nodeFs.mkdirSync(nodePath.dirname(outputFile.path), {recursive: true});
                    nodeFs.writeFileSync(outputFile.path, outputFile.text.replaceAll(search, replace));
                }
            })
        },
    }
]
```


</p>
</details> 

@evanw could you please tell us if you need help testing or discussing https://github.com/evanw/esbuild/pull/3914 or https://github.com/evanw/esbuild/pull/4391
